### PR TITLE
Use request.characterEncoding prior to default encoding

### PR DESCRIPTION
**Pull request by marcosbarbero**
Zuul is using ISO-8859-1 as default character encoding making requests fail when it's expected any other encoding. The problem occurs at URI encoding and queryParams encoding.

Related to this [post from stackoverflow](http://stackoverflow.com/questions/36475845/netflix-zuul-query-string-encoding)

### CODE REVIEW

File: spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/SampleZuulProxyApplicationTests.java [hunk 1]
```diff
@@ -146,7 +146,7 @@ public void simpleHostRouteWithEncodedQuery() {
 				"http://localhost:" + this.port + "/self/query?foo={foo}", HttpMethod.GET,
 				new HttpEntity<>((Void) null), String.class, "weird#chars");
 		assertEquals(HttpStatus.OK, result.getStatusCode());
-		assertEquals("/query?foo=weird#chars", result.getBody());
+		assertEquals("/query?foo=weird%23chars", result.getBody());
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/SampleZuulProxyApplicationTests.java]: I'm worried that this test changed. The `#` char has meaning in a url, if it's changed, I'm not sure I can accept this.

Comment (marcosbarbero) [inline spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/SampleZuulProxyApplicationTests.java]: @spencergibb you're right the '#' is a gen-delims as specified at RFC, I didn't payed attention on that. 
Anyhow, the URI encoding is a real problem. I'll fix that PR.

**Comment by marcosbarbero** [comment]
@spencergibb the encoding URI is a real problem. I'll give you a scenario.
1. Sending a request to `/zuul/any-sp%C3%A9cial-char/action` with `UTF-8` encoding
2. `ProxyRequestHelper` class will receive it like `/zuul/any-spécial-char/action`
3. `ProxyRequestHelper` class will encode it like `/zuul/any-sp%E9cial-char/action` with  `ISO-8859-1` encoding
4. Zuul routes the request to the final service and it cannot read the request because it is expecting `UTF-8` instead of `ISO-8859-1`

---
My approach here is to check the incoming request encoding and try to use it to encode de URI instead of always use the default value.

**Comment by spencergibb** [comment]
@marcosbarbero I'm not debating that it is not a real problem. Some tests would be helpful.

**Comment by marcosbarbero** [comment]
@spencergibb I'm sorry if it sounded rude, I'd not intention to say you were debating.
I'll try to provide some tests in few days.

**Comment by spencergibb** [comment]
@marcosbarbero NP! Looking forward to the tests.

**Comment by marcosbarbero** [comment]
@spencergibb I've just added few tests. Hope it helps, if need some more scenarios drop me a message :)

**Comment by spencergibb** [comment]
Nice!

**Comment by spencergibb** [comment]
Merged via 1c532d96375485b8d551cbbe766507bacfaafc26

**Pull request closed without merging**